File: Day1_PointerBasics_Go/hard/hard_02.go
// FILE: hard_02.go
// TOPIC: Pointer Basics - Day 1 (Go)
// DIFFICULTY: Hard
//
// TASK: Pass a pointer to a function to modify the caller's variable
// ------------------------------------------------------------------
// Go is pass-by-value. To let a function modify the caller's variable,
// pass a POINTER to it.
//
// 1. Write:  func doubleIt(ptr *int)  — multiplies *ptr by 2.
// 2. Write:  func addFive(ptr *int)   — adds 5 to *ptr.
// 3. In main():
//      a) Declare n := 10
//      b) Call doubleIt(&n). Print n.
//      c) Call addFive(&n).  Print n.
//      d) Call doubleIt(&n). Print n.
//
// EXPECTED OUTPUT:
//   After doubleIt: 20
//   After addFive : 25
//   After doubleIt: 50

package main

import "fmt"

// TODO: Write doubleIt and addFive here.

func main() {
	// TODO: Write your solution here.

}

File: Day1_PointerBasics/hard/hard_02.cpp
/*
 * FILE: hard_02.cpp
 * TOPIC: Pointer Basics - Day 1
 * DIFFICULTY: Hard
 *
 * TASK: Null pointers and detecting an invalid pointer
 * ----------------------------------------------------
 * A null pointer holds the address 0 (nullptr in modern C++).
 * Dereferencing it is undefined behaviour - your program will crash.
 * This exercise teaches you to GUARD against that.
 *
 * 1. Declare an int pointer `p` and initialise it to nullptr.
 * 2. Write an if-statement that checks whether `p` is null.
 *    If null: print "Pointer is null - cannot dereference safely."
 *    If not null: print the value it points to.
 * 3. Now assign `p` to point to a local int variable `x = 77`.
 * 4. Run the same if-check again and print the value this time.
 *
 * EXPECTED OUTPUT:
 *   Pointer is null - cannot dereference safely.
 *   Pointer is valid. Value = 77
 *
 * BONUS QUESTION (think about it, no code needed):
 *   Why is it important to always initialise pointers to nullptr
 *   rather than leaving them uninitialised?
 */

#include <iostream>
using namespace std;

int main() {
    // TODO: Write your solution here.
    int* p;
    int x = 77;
    int num = 0;

    while (num < 2) {
        if (num == 0) {
            if (p == nullptr) {
                cout << "Pointer is null - cannot dereference safely." << "\n";
            };
        };
        if (num == 1) {
            p = &x;
            if (p != nullptr) {
                cout << "Pointer is valid. Value = " << *p << "\n";
            };
        };
        num++;
    };

    return 0;
}
